apis/stacks/v1alpha1/stackdefinition.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// StackDefinitionSpec defines the desired state of StackDefinition
+type StackDefinitionSpec struct {
+	AppMetadataSpec `json:",inline"`
+	CRDs            CRDList                 `json:"customresourcedefinitions,omitempty"`
+	Controller      ControllerSpec          `json:"controller,omitempty"`
+	Permissions     PermissionsSpec         `json:"permissions,omitempty"`
+	Behavior        StackDefinitionBehavior `json:"behavior,omitempty"`
+}
+
+// StackDefinitionSource is the stack image which this stack configuration is from.
+// In the future, other source types may be supported, such as a URL.
+type StackDefinitionSource struct {
+	// a container image id
+	Image string `json:"image,omitempty"`
+}
+
+// StackDefinitionBehavior specifies the behavior for the stack, if the stack has behaviors instead of a controller
+type StackDefinitionBehavior struct {
+	CRD    StackDefinitionBehaviorCRD       `json:"crd,omitempty"`
+	Engine StackResourceEngineConfiguration `json:"engine,omitempty"`
+	// Theoretically, source and engine could be specified at a per-hook level
+	// as well.
+	Source    StackDefinitionSource              `json:"source,omitempty"`
+	Reconcile []StackDefinitionHookConfiguration `json:"reconcile,omitempty"`
```

Comment:
Not necessarily, but, we do have some internal use-cases which use multiple separate helm charts for managing a single service, so the most natural pattern would be to have multiple hooks for reconcile. And that seems pretty reasonable to me, so I'm thinking it makes sense to leave it as an array.

Reply:
I was thinking a bit more about this as I was updating the types, and it seems like we can just make this a single hook. There are two reasons. The first reason is that if people can specify multiple hooks, it'd make sense for them to also want to specify different sources for different hooks, which we weren't planning on supporting right away. The second reason is the idea of template stacks is to allow people to easily compose other pieces, so it seems like it would make sense to ask them to compose them using the "normal" composition approach (which is by using a resource engine such as helm2 or kustomize to define a composite resource set). Making this simplification now also fits with our overarching approach of satisfying the simple, expected use-case first and then making things more complex as needed, which I agree with.